pyccel/stdlib/ndarrays/ndarrays.c

```diff
@@ -373,3 +373,240 @@ int64_t     *numpy_to_ndarray_shape(int64_t *np_shape, int nd)
     return nd_shape;
 
 }
+
+/*
+** sum of ndarray
+*/
+
+static int64_t     get_index_from_array(t_ndarray arr, int64_t *tmp)
+{
+    switch (arr.nd) {
+        case 1:
+            return GET_INDEX(arr, tmp[0]);
+        case 2:
+            return GET_INDEX(arr, tmp[0], tmp[1]);
+        case 3:
+            return GET_INDEX(arr, tmp[0], tmp[1], tmp[2]);
+        case 4:
+            return GET_INDEX(arr, tmp[0], tmp[1], tmp[2], tmp[3]);
+        case 5:
+            return GET_INDEX(arr, tmp[0], tmp[1], tmp[2], tmp[3], tmp[4]);
+        case 6:
+            return GET_INDEX(arr, tmp[0], tmp[1], tmp[2], tmp[3], tmp[4], tmp[5]);
+        case 7:
+            return GET_INDEX(arr, tmp[0], tmp[1], tmp[2], tmp[3], tmp[4], tmp[5], tmp[6]);
+        case 8:
+            return GET_INDEX(arr, tmp[0], tmp[1], tmp[2], tmp[3], tmp[4], tmp[5], tmp[6], tmp[7]);
+        case 9:
+            return GET_INDEX(arr, tmp[0], tmp[1], tmp[2], tmp[3], tmp[4], tmp[5], tmp[6], tmp[7], tmp[8]);
+        case 10:
+            return GET_INDEX(arr, tmp[0], tmp[1], tmp[2], tmp[3], tmp[4], tmp[5], tmp[6], tmp[7], tmp[8], tmp[9]);
+    }
+}

```

Comment:
This raises a warning:
```
  /Users/runner/hostedtoolcache/Python/3.10.9/x64/lib/python3.10/site-packages/pyccel/codegen/compiling/compilers.py:431: UserWarning: /Users/runner/work/pyccel/pyccel/tests/internal/scripts/openmp/__pyccel__/ndarrays/ndarrays.c:405:1: warning: non-void function does not return a value in all control paths [-Wreturn-type]
```

As per this:
https://github.com/pyccel/pyccel/blob/f47756a0eb0c9b6da96037ff077df0e06e84b9c3/pyccel/stdlib/ndarrays/ndarrays.h#L31-L45
Pyccel can handle up to 15 dimensions. The default should raise an error before so you could use something like:
```
default:
    return -1;
```

That said, I don't know if macro+switch is more efficient than just using a for loop to calculate the index.

Reply:
Sorry, I will add the default return and i will go up to 15 dimensions. The problem with using for loops is that we only know the number of for loops at runtime. How can we generate such a loop dynamically in C?

Reply:
I don't think we need multiple loops here:
```
static int64_t     get_index_from_array(t_ndarray arr, int64_t *tmp)
{
    int64_t idx = 0;
    for (int dim(0); dim<arr.nd; ++dim) {
        idx += arr.strides[dim] * (tmp[dim]);
    }
    return idx;
}
```

Reply:
That's much better! thank you!